### ISSUE TITLE
```ClassVar``` does not work properly with ```staticmethod```

### ISSUE THREAD

**Issue by daler-sz** [python/mypy#16947]
**Bug Report**

```ClassVar``` deos not work properly with ```staticmethod```. 
```staticmethod``` descriptor indicates that the callable member of class would be called as it is, without ```self``` parameter. But ```mypy``` and ```pyright``` both can not resolve this.

**To Reproduce**

```python
from typing import ClassVar, Callable


def func(x: int) -> None:
    return None


class A:
    f: ClassVar[Callable[[int], None]] = staticmethod(func)


A().f(2)
```
Playground: https://mypy-play.net/?mypy=latest&python=3.12&gist=f758b51682a7148792aacf59149b9cf6

**Expected Behavior**

No error expected.

**Actual Behavior**

```
error: Invalid self argument "A" to attribute function "f" with type "Callable[[int], None]"  [misc]
error: Too many arguments  [call-arg]
```
Typecheckers think that ```self``` parameter passed to function, ignoring the fact that this is ```staticmethod``` descriptor.

Yes, this is false negative output. But there is also false positive case: https://mypy-play.net/?mypy=latest&python=3.12&gist=c2b575b4e6058a989701f5fa54dca052

**Your Environment**

- Mypy version used: 1.8.0
- Python version used: 3.12

**Comment by erictraut** [comment]
There's currently no way in the type system to differentiate between a `Callable` that should be treated as a static method versus one that should be treated as an instance method or class method. When a type checker sees an annotation `Callable[[int], None]`, it needs to make an assumption about which of these it is. Both mypy and pyright (reasonably) assume it's an instance variable in the code above. I don't see a way to fix this without augmenting the typing spec, perhaps adding a new special form to the `typing` module that differentiates between the different types of callables. I don't know if there's going to be much appetite for such a change given that this is a rare use case, but you're welcome to start a conversation in the [typing forum](https://discuss.python.org/c/typing/32) if you'd like.

In the meantime, the approach I recommend is to use a callback protocol to declare the type of `f`.

```python
from typing import ClassVar, Protocol

class MyCallback(Protocol):
    @staticmethod
    def __call__(x: int, /) -> None: ...

def func(x: int) -> None:
    return None

class A:
    f: ClassVar[MyCallback] = staticmethod(func)

A().f(2)
```

The above code type checks without errors in pyright, but it runs into another bug in mypy.
